### ISSUE TITLE
HMM, Multiple Binary variables, and from_sample

### ISSUE THREAD

**Issue by rzu512** [jmschrei/pomegranate#477]
If a hidden variable H is True, then observable variable A1 has 30% chance to be True, and observable variable A2 has 60% chance to be True. I am not sure if A1 and A2 are independent. But I do not know their relationships. There are 30,000 observable variables.

I think I want a JointProbabilityTable. But how to use that with from_sample?
Test:
```python
import pomegranate as pg
import numpy as np

n = 4
d = 1000
x = np.random.randint(2, size=(n, d))
x[3] = x[1] + 2 * x[2]
X = [x]

model = pg.HiddenMarkovModel.from_samples(
    pg.JointProbabilityTable,
    n_components=2, X=X,
    verbose=True)
```

Error:
```python
Traceback (most recent call last):
  File "<...>/hmm.py", line 13, in <module>
    verbose=True)
  File "pomegranate/hmm.pyx", line 3594, in pomegranate.hmm.HiddenMarkovModel.from_samples
  File "pomegranate/distributions.pyx", line 3035, in pomegranate.distributions.JointProbabilityTable.from_samples
TypeError: from_samples() takes at least 2 positional arguments (1 given)
```
Version:
```
>>> import pomegranate as pg
>>> pg.__version__
'0.9.0'
```

By the way, stackoverflow does not have a pomegranate tag.
I tried to search in closed issues to look for past answers on similar problems.
Is it ok to ask question here?

**Comment by jmschrei** [comment]
Yes, please do ask questions here. I don't really monitor stack overflow.

As for your question-- unless there's a sequential element to your data, it sounds like you want a Bayesian network. Your network would have three variables, H, A1, and A2. You would have arrows going from H to A1 and A2, and pass in a ConditionalProbabilityTable as the distribution for A1 and A2. H would be modeled with a Discrete Distribution. This is saying that A1 and A2 are ~conditionally~ independent given that you know H, but if you don't know H then there is some covariance between them.

Is there a sequential element to your data? Can you give a minimal example of data with the same shape?

**Comment by rzu512** [comment]
How to use a `JointProbabilityTable` with "from_sample" for a HMM? I tried in the code that I posted, but the code does not work. Is it possible to fix that?

I have to get a multi-variate hidden Markov model. The input are 2D Boolean arrays representing Boolean variables that changes over time. That would be something like in the code I posted above:

`x = np.random.randint(2, size=(n, d), dtype=bool))`

I have tried 'MultiVariateGaussian` and it cannot work with "from_sample" because the "leading minor is not positive define" problem. I suspect it is because covariance matrices are positive semidefinite instead of positive definite ([link](https://mathoverflow.net/questions/155147/cholesky-decomposition-of-a-positive-semi-de%EF%AC%81nite)).

**Comment by jmschrei** [comment]
Hi @rzu512 . The issue you mention with MultivariateGaussian should be fixed in pomegranate v0.10.0. I'd suggest you download the latest code (available on PyPI or cloned from here) and see if it works. 

I don't think that you want JointProbabilityTable. For boolean variables that would have 2**d elements in it. You probably want to use IndependentComponentsDistribution with Bernoulli variables, which would only have d parameters. Also, you can't do from_samples with JointProbabilityTable right now.

If you have multivariate data, pomegranate expects a list of sequences, where each sequence is a 2D numpy array. If it's a single long sequence the sequence still needs to be wrapped in a list, even if it's the only element in the list.

**Comment by agcala** [comment]
Pomegranate seems to be a powerful and useful package yet is hard to use:
a) Because the statistical knowledge required is above the average user.
b) The documentation is skimpy and examples are scarce. 
Regardless of which is the right distribution for the OP to employ, the question remains of how to use the JointProbabilityTable distribution and which if any are the arguments expected. There is nothing in the documents to suggest one should provide any additional arguments.

**Comment by jmschrei** [comment]
I'm sorry you're having difficulties. I agree that there are several areas that can be better fleshed out. 

Ultimately, the reason behind these gaps is because pomegranate has always been a side project for me while I work on my PhD research (which is unfortunately totally unrelated to pomegranate) or other packages (which are more related to my research). When I have free time to work on pomegranate, I need to choose between more features, more tutorials/examples, more documentation, and debugging / adding in unit tests. 

Further, while the many contributors to the project are invaluable and have certainly made pomegranate better, generally these additions focus on new features or fixes for bugs rather than additional unit tests and examples of how to use their proposed features, leaving me with the work of adding those in as well. 

If you're unhappy with the amount of documentation or examples, I'd strongly encourage you to consider contributing to the project so that it can become even better. I'm happy to accept short notebooks on any topic.
